src/CodeAnalysis/CSharp/MetaAnalyzers/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,113 @@
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class CSharpImmutableObjectMethodAnalyzer : DiagnosticAnalyzer
+    {
+        // Each analyzer needs a public id to identify each DiagnosticDescriptor and subsequently fix diagnostics in CodeFixProvider.cs
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationTitle), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationMessage), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationDescription), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+
+        public static DiagnosticDescriptor DoNotIgnoreReturnValueDiagnosticRule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocation,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.AnalyzerCorrectness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(DoNotIgnoreReturnValueDiagnosticRule);
+            }
+        }
+
+        private static readonly string s_solutionFullName = @"Microsoft.CodeAnalysis.Solution";
+        private static readonly string s_projectFullName = @"Microsoft.CodeAnalysis.Project";
+        private static readonly string s_documentFullName = @"Microsoft.CodeAnalysis.Document";
+        private static readonly string s_syntaxNodeFullName = @"Microsoft.CodeAnalysis.SyntaxNode";
+
+        // A list of known immutable object names
+        private static ImmutableArray<string> s_immutableObjectNames
+        {
+            get
+            {
+                return ImmutableArray.Create(s_solutionFullName, s_projectFullName, s_documentFullName, s_syntaxNodeFullName);
+            }
+        }
+
+        private static readonly string s_Add = "Add";
+        private static readonly string s_Remove = "Remove";
+        private static readonly string s_Replace = "Replace";
+        private static readonly string s_With = "With";
+
+        private static ImmutableArray<string> s_immutableMethodNames 
```

Comment:
I left out `Get` methods because I don't think they're as likely to confuse users as the name doesn't suggest a possibility for mutation the same way Add, Remove, Replace and With do.

I've also left out the `Workspace` for now because the `AdhocWorkspace` isn't immutable. If you guys think it's worth special-casing so we can include workspace, I'll add it.